src/components/legend.ts

```diff
@@ -159,70 +438,68 @@ namespace Plottable.Components {
+          table.addRow(row);
+        }
 
-      let rows = this._packRows(availableWidthForEntries, entryNames, entryLengths);
+        const symbolColumn = new LegendColumn(textHeight, textHeight, entryName);
+        row.addColumn(symbolColumn);
 
-      let rowsAvailable = Math.floor((availableHeight - 2 * this._padding) / textHeight);
-      if (rowsAvailable !== rowsAvailable) { // rowsAvailable can be NaN if this.textHeight = 0
-        rowsAvailable = 0;
-      }
+        // the space consumed by the name field is the minimum of the space available in the table
+        // and the actual width consumed by the name
+        availableWidth = row.getWidthAvailable();
+        const usedNameWidth = Math.min(availableWidth, unwrappedNameWidth);
 
-      return {
-        textHeight: textHeight,
-        entryLengths: entryLengths,
-        untruncatedEntryLengths: untruncatedEntryLengths,
-        rows: rows,
-        numRowsToDraw: Math.max(Math.min(rowsAvailable, rows.length), 0),
-      };
-    }
-
-    public requestedSpace(offeredWidth: number, offeredHeight: number): SpaceRequest {
-      let estimatedLayout = this._calculateLayoutInfo(offeredWidth, offeredHeight);
+        this._wrapper.maxLines(this.maxLinesPerEntry());
+        let numberOfRows = this._wrapper.wrap(formattedName, this._measurer, usedNameWidth).noLines;
 
-      let untruncatedRowLengths = estimatedLayout.rows.map((row) => {
-        return d3.sum(row, (entry) => estimatedLayout.untruncatedEntryLengths.get(entry));
+        let nameColumnHeight = numberOfRows * textHeight;
+        const nameColumn = new LegendColumn(usedNameWidth, nameColumnHeight, entryName);
+        row.addColumn(nameColumn);
       });
-      let longestUntruncatedRowLength = Utils.Math.max(untruncatedRowLengths, 0);
 
-      return {
-        minWidth: this._padding + longestUntruncatedRowLength,
-        minHeight: estimatedLayout.rows.length * estimatedLayout.textHeight + 2 * this._padding,
-      };
+      return table;
     }
 
-    private _packRows(availableWidth: number, entries: string[], entryLengths: d3.Map<number>) {
-      let rows: string[][] = [];
-      let currentRow: string[] = [];
-      let spaceLeft = availableWidth;
-      entries.forEach((e: string) => {
-        let entryLength = entryLengths.get(e);
-        if (entryLength > spaceLeft || currentRow.length === this._maxEntriesPerRow) {
-          rows.push(currentRow);
-          currentRow = [];
-          spaceLeft = availableWidth;
-        }
-        currentRow.push(e);
-        spaceLeft -= entryLength;
-      });
+    public requestedSpace(offeredWidth: number, offeredHeight: number): SpaceRequest {
+      // if max width is set, the table is guaranteed to be at most maxWidth wide.
+      // if max width is not set, the table will be as wide as the longest untruncated row
+      const table = this._buildLegendTable(Utils.Math.min([this.maxWidth(), Infinity], Infinity), offeredHeight);
```

Comment:
extra infinity is the default value that will be returned if, say, `this.maxWidth` returns null .